### ISSUE TITLE
Touchscreen support

### ISSUE THREAD

**Issue by avodonosov** [jmoenig/Snap#2893]
For example, it is impossible currently to choose fill color in the Vector Paint Editor on touch devices - the fill color can only be set with right mouse click.

**Comment by Simon-Mong** [comment]
Oh,I have that in my home.  But only compatible with 5.x versions of Snap! I will  try to do it with new version of Snap!

**Comment by WarpedWartWars** [comment]
Yes, long-hold should act like a right-click in all cases. (Except those in which the programmer *explicitly* states the two are different.)

**Comment by avodonosov** [comment]
Not nesesserily through a long-hold, I would be satisfied with watever way. In Scratch the costume editor has a separate color dropdown for the fill color.

BTW, my story is that I've created a game in Scratch using my tablet. Further imrovements of the are blocked by the language limitations. I've turned to Snap!, but it does not support touch screen.

**Comment by WarpedWartWars** [comment]
> does not support touch screen

Well, it partially *does*, just not completely.
> Not nesesserily through a long-hold

Long-hold is the standard right-click mechanism on mobile devices.
> language limitations

~~What language?~~ Oh, you mean "limitations of Scratch", not of a *spoken* language.

**Comment by WarpedWartWars** [comment]
> > > does not support touch screen
> 
> > Well, it partially does, just not completely.
> 
> Well, specifically the fill color problem blocks me from drawing a consume in vector editor. Consumes are so basic part of the system, if touchscreen not supported here I am afraid to move forward with workarounds for that specific problem (raster image instead?).

I'm not sure what you'll have to do until then. Maybe make them in an external vector editor?

**Comment by WarpedWartWars** [comment]
Sorry for not replying earlier, but could you try two-finger tap? That might work.

**Comment by avodonosov** [comment]
@WarpedWartWars , thanks for the reply, but unfortonately two-finger tap does not work. And it would not je appropiate, because the point of the click should choose the color, and the two finger tap does not have exact point (it has two points).

**Comment by WarpedWartWars** [comment]
> (it has two points)

The point halfway between then?

> but unfortonately two-finger tap does not work

I might be able to implement it and make a pull request.

**Comment by avodonosov** [comment]
@WarpedWartWars , I am not entierly sure about usability of this approach, only trying working solution can show. But it will be better than nothing.

**Comment by jmoenig** [comment]
Folks, I've fixed this for the upcoming version. Touch-hold now lets you select the fill-color in the vector editor on iPads etc. Please give it a try at https://snap.berkeley.edu/versions/dev/snap.html and let me know if not.
Thanks for reminding me about this issue, btw!

**Comment by avodonosov** [comment]
@jmoenig, what I observe in Chrome on Android
- long-hold is unable to change the fill color:
  - on the "finger down" the edge-color changes
  - neither holding the finger, nor taking the finger up after the hold has any effect
- two-finger tap can be used as a kind of not very convenient workaround:
  - the "fingers down" canges the edge-color according to the first finger landed
  - the "fingers up" changes the fill-color according to the last finger taken up.
  - as you see, the inconvenience here is that the user can not change fill-color without changing the edge-color at the same timed

**Comment by avodonosov** [comment]
A robust and clean, in my opinion, way to fix that, would be to turn the labels "Edge color" and "Fill color" ino a radio button. And apply the current edge color logic to fill color, if the radio button is switched to the "Fill color".

An advantage of this approach is that the fill color will keep changing as I move finger - the same way as it works for the edge color now.

It would also be more convenient if the whole pane of the currently selected edge and fill colors was placed above the color chosing field, so that my hand does cover the pane while I touch the field, so that it is convenient to observe the changes of the selected value as I move the finger.

**Comment by avodonosov** [comment]
One more problem on touch screens: in the vector mode of costume editor, how do I delete individual ojjects from the drawing?

There is a selection tool, prbably on desktop pressing the Delete key removes the selected elements. But on a touchscreen I don't see a way to delete the selected elements.

**Comment by WarpedWartWars** [comment]
> in the vector mode of costume editor

Someone (@brianharvey) said the vector editor would be remade in the future. (I'm not sure when, though. Probably after v7.)
